realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -0,0 +1,351 @@
+        }
+        env->SetLongArrayRegion(version_data, 0, 2, version_array);
+
+        return version_data;
+    } CATCH_STD ()
+
+    return NULL;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_SharedRealm_nativeIsClosed(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    return static_cast<jboolean>(shared_realm->is_closed());
+}
+
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetTable(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                  jstring table_name) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    try {
+        JStringAccessor name(env, table_name); // throws
+        auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+        if (!shared_realm->read_group().has_table(name) && !shared_realm->is_in_transaction()) {
+            std::ostringstream ss;
+            ss << "Table " << name << " doesn't exist and the shared realm is not in transaction.";
+            ThrowException(env, IllegalState, ss.str());
+            return static_cast<jlong>(NULL);
+        }
+        Table* pTable = LangBindHelper::get_or_add_table(shared_realm->read_group(), name);
+        return reinterpret_cast<jlong>(pTable);
+    } CATCH_STD()
+
+    return static_cast<jlong>(NULL);
+}
+
+JNIEXPORT jstring JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetTableName(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                      jint index) {
+
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        return to_jstring(env, shared_realm->read_group().get_table_name(S(index)));
+    } CATCH_STD()
+    return NULL;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_SharedRealm_nativeHasTable(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                  jstring table_name) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        JStringAccessor name(env, table_name);
+        return static_cast<jboolean >(shared_realm->read_group().has_table(name));
+    } CATCH_STD()
+    return static_cast<jboolean >(false);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeRenameTable(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                     jstring old_table_name, jstring new_table_name) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        JStringAccessor old_name(env, old_table_name);
+        if (!shared_realm->is_in_transaction()) {
+            std::ostringstream ss;
+            ss << "Table " << old_name << " cannot be removed when the shared realm is not in transaction.";
```

Comment:
It is more like a internal exception. We could change it to RealmObject, but what if we internally calling this method to remove a pk table? The the message will be quite confusing ... So, any better term than Table or RealmObject?